ste/pageRangeOptimizer.go

```diff
@@ -0,0 +1,117 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ste
+
+import (
+	"context"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/pageblob"
+	"github.com/Azure/azure-storage-azcopy/v10/common"
+	"strings"
+)
+
+// isolate the logic to fetch page ranges for a page blob, and check whether a given range has data
+// for two purposes:
+//  1. capture the necessary info to do so, so that fetchPages can be invoked anywhere
+//  2. open to extending the logic, which could be re-used for both download and s2s scenarios
+type pageRangeOptimizer struct {
+	srcPageBlobClient *pageblob.Client
+	ctx               context.Context
+	srcPageList       *pageblob.PageList // nil if src is not a page blob, or it was not possible to get a response
+}
+
+func newPageRangeOptimizer(srcPageBlobClient *pageblob.Client, ctx context.Context) *pageRangeOptimizer {
+	return &pageRangeOptimizer{srcPageBlobClient: srcPageBlobClient, ctx: ctx}
+}
+
+// withNoRetryForBlob returns a context that contains a marker to say we don't want any retries to happen
+// Is only implemented for blob pipelines at present
+func withNoRetryForBlob(ctx context.Context) context.Context {
+	return policy.WithRetryOptions(ctx, policy.RetryOptions{MaxRetries: -1})
+}
+
+func (p *pageRangeOptimizer) fetchPages() {
+	// don't fetch page blob list if optimizations are not desired,
+	// the lack of page list indicates that there's data everywhere
+	if !strings.EqualFold(common.GetLifecycleMgr().GetEnvironmentVariable(
+		common.EEnvironmentVariable.OptimizeSparsePageBlobTransfers()), "true") {
+		return
+	}
+
+	// according to the REST API documentation:
+	// in a highly fragmented page blob with a large number of writes,
+	// a Get Page Ranges request can fail due to an internal server timeout.
+	// thus, if the page blob is not sparse, it's ok for it to fail
+	// TODO follow up with the service folks to confirm the scale at which the timeouts occur
+	// TODO perhaps we need to add more logic here to optimize for more cases
+	limitedContext := withNoRetryForBlob(p.ctx) // we don't want retries here. If it doesn't work the first time, we don't want to chew up (lots) time retrying
+	pager := p.srcPageBlobClient.NewGetPageRangesPager(&pageblob.GetPageRangesOptions{})
```

Comment:
Changed this from nil to empty options bag due to https://github.com/Azure/azure-sdk-for-go/issues/22356